okpt/io/config/validator.py

```diff
@@ -0,0 +1,109 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+
+# Modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+"""Provides validation for various config files.
+
+Functions:
+    validate(): Validate the tool configuration.
+
+Exceptions:
+    ConfigurationError: An error in the tool configuration.
+
+"""
+
+import os
+from io import TextIOWrapper
+from typing import Any, Dict
+
+import cerberus
+
+from okpt.io import reader
+from okpt.io.config import parser
+
+_TOOL, _OPENSEARCH = ['tool', 'opensearch']
+
+
+class ConfigurationError(Exception):
+    """Exception raised for errors in the tool configuration.
+
+    Attributes:
+        message -- explanation of the error
+    """
+    def __init__(self, message: str):
+        self.message = message
+
+
+class _Validator():
+    """Validator for various configuration schemas.
+
+    Attributes:
+        validator: Cerberus validator for a particular schema
+        errors: Cerberus validation errors (if any are found during validation)
+    """
+    def __init__(self, schema_name: str):
+        self.validator = self._get_validator_from_schema_name(schema_name)
+        self.errors = ''
+
+    def _get_validator_from_schema_name(self, schema_name: str):
+        """Get the corresponding Cerberus validator from a schema name."""
+        curr_file_dir = os.path.dirname(os.path.abspath(__file__))
+        schemas_dir = os.path.join(curr_file_dir, 'schemas')
+        schema_file_path = os.path.join(schemas_dir, f'{schema_name}.yml')
+        schema_file_obj = reader.get_read_file_obj(schema_file_path)
+        schema_obj = parser.parse_yaml(schema_file_obj)
+        return cerberus.Validator(schema_obj)
+
+    def validate(self, config_obj: Dict[str, Any]):
+        """Validate the configuration obj against the class schema."""
+        if not self.validator.validate(config_obj):
+            self.errors = self.validator.errors
+            return False
+        return True
+
+
+def validate(tool_config_file_obj: TextIOWrapper):
+    """Validate the configurations of the tool.
+
+    Args:
+        tool_config_file_obj: File object of the tool configuration file.
+
+    Returns:
+        List with the tool configuration, service configuration, and index settings (if possible).
+
+    Raises:
+        ConfigurationError: If there is an error in the user configuration.
+    """
+
+    # validate tool config
+    tool_config_validator = _Validator(_TOOL)
+    tool_config_obj = parser.parse_yaml(tool_config_file_obj)
+    is_tool_config_valid = tool_config_validator.validate(tool_config_obj)
+
+    if is_tool_config_valid:
```

Comment:
Please don't have logic nested . It is hard to read. I would try to flip conditions and exit or raise. Same applicable for else as well.

Reply:
I've flattened out the validation logic, let me know if it should be further refactored.